pipenv/core.py

```diff
@@ -733,14 +733,11 @@ def batch_install(deps_list, procs, failed_deps_queue,
                 trusted_hosts=trusted_hosts,
                 extra_indexes=extra_indexes
             )
-            if dep.is_vcs or dep.editable:
-                c.block()
-            if procs.qsize() < nprocs:
-                c.dep = dep
-                procs.put(c)
+            c.dep = dep
 
-            if procs.full() or procs.qsize() == len(deps_list):
-                _cleanup_procs(procs, not blocking, failed_deps_queue, retry=retry)
+            procs.put(c)
+            if procs.full():
```

Comment:
Assuming we're talking about `if procs.full() or procs.qsize() == len(deps_list):`...

This one's a bit awkward. I've failed to find the reason for this check by myself and also can't parse your explanation. I believe:

 *  This function must force the cleanup when the queue is full because running new processes would violate the `nprocs` limit.
 * Otherwise it doesn't strictly need to clean up ever because the caller will clean up regardless.
 * The condition seems to only be true at the last iteration of the loop when all dependencies have fit in the queue at once, but I don't see the significance.

Reply:
it is only true in the last run of a `concurrent=True` cycle through the loop and I believe there was some race condition where not all processes were cleaned up during one pass if it didn't saturate the queue...

It is possible it isn't needed, however removing an additional safeguard in a bugfix PR where you are mixing in refactoring is my primary issue. There is a good chance it is like this for a reason, even if that reason is not documented